crmsh/join_lock.py

```diff
@@ -0,0 +1,167 @@
+            elif re.search("ssh: connect to .*: Connection timed out", err):
+                raise TimeoutError(err)
+            else:
+                raise RuntimeError(err)
+        return rc, out
+
+    def _get_last_timestamp(self):
+        """
+        Tail the last line in lock file
+        """
+        result = {}
+        rc, out = self._run(self.TAIL_CMD)
+        if rc == 0 and out:
+            try:
+                locked_node, timestamp = out.split()
+            except ValueError:
+                raise ValueError("Invalid format for {}(should be \"node timestamp\")".format(self.JOIN_LOCK_FILE))
+            else:
+                result['node'] = locked_node
+                result['timestamp'] = timestamp
+        return result
+
+    def _remove_lock_dir(self):
+        """
+        Remove lock directory
+        """
+        self._run(self.RM_CMD)
+
+    def _create_lock_dir(self):
+        """
+        Create lock directory, mkdir command was atomic
+        """
+        try:
+            self._run(self.MKDIR_CMD)
+        except FileExistsError:
+            return False
+        return True
+
+    def _lock_or_wait(self):
+        """
+        Wait 15 seconds while detect target node has lock directory
+        """
+        file_not_found_once = False
+        warned_once = False
+        origin_timestamp = None
+
+        while not self._create_lock_dir():
+
+            try:
+                result = self._get_last_timestamp()
+            # consider lock file might not be created yet
+            except FileNotFoundError:
+                if file_not_found_once:
+                    self._remove_lock_dir()
+                    break
+                else:
+                    file_not_found_once = True
+                    time.sleep(self.DETECT_INTERVAL)
+                    continue
+
+            locked_node, timestamp = result.values()
+            if origin_timestamp:
+                res = int(timestamp) - int(origin_timestamp)
+                if res < self.UPDATE_INTERVAL:
+                    self._remove_lock_dir()
+                    break
+            origin_timestamp = timestamp
+
+            if not warned_once:
+                warned_once = True
+                bootstrap.warn("Node {} still joining, wait...".format(locked_node))
+
+            time.sleep(self.DETECT_INTERVAL)
+
+    def _update_timestamp(self):
+        """
+        Update timestamp(10s interval) in lock file on target node, format is:
+            nodename timestamp
+        Any exceptions will lead to kill the whole join process
+        """
+        while not self.thread_stop_event.is_set():
+            try:
+                self._run(self.UPDATE_CMD)
+            except Exception as err:
+                bootstrap.print_error_msg(str(err))
+                pid = os.getpid()
+                os.kill(pid, signal.SIGQUIT)
+                break
+            time.sleep(self.UPDATE_INTERVAL)
```

Comment:
`self._run` with update timestamp shell command might raise any exceptions(maybe include ssh timed out)
My thinking was to terminate the whole join process.
But it's hard to catch child thread's exception in main thread without join(which will block the join process).
So I use signal to kill the join pid
Seems a bit complex, but it works